src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -130,8 +118,8 @@ public SerializableOptionSet WithLanguages(ImmutableHashSet<string> languages)
         }
 
         private bool ShouldSerialize(OptionKey optionKey)
-            => _serializableOptions.Contains(optionKey.Option) &&
-               (!optionKey.Option.IsPerLanguage || _languages.Contains(optionKey.Language!));
+            => _serializableOptionValues.ContainsKey(optionKey) &&
```

Comment:
I am wondering if this breaks the logic in `WithChangedOption` here: https://github.com/dotnet/roslyn/blob/633346af571d640eeacb2e2fc724f5d25ed20faa/src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs#L153

If `WithChangedOption` is invoked with a completely new serializable option that is not part of the current serializable option set (and hence not part of `_serializableOptionValues`), prior logic would have added the option and value to `_serializableOptionValues`, but the new logic would end up considering it as a non-serializable option and add it to `_workspaceOptionSet` instead.

Reply:
yeah... i'm not sure what semantics we want here.  so there is some totally new option... and it's for what?  some langauge we've never loaded?  What *do* we want to do there?

I mean the serializable options are effectively on the C#/VB options (we explicitly check that).  So what is a case of a C#/VB option that we ahve that isn't already loaded up front and which we get a 'new serializable option' for?

Reply:
> So there is some totally new option... and it's for what? some langauge we've never loaded?

From what I recall from past debugging, not all serializable option values are loaded upfront. Some of them get loaded in delayed fashion. And yes once we add a project with a new language, it will certainly lead to new language specific options getting loaded. You may want to confirm by debugging through.

Reply:
> And yes once we add a project with a new language, it will certainly lead to new language specific options getting loaded.

Yup.  That part is fine.

> From what I recall from past debugging, not all serializable option values are loaded upfront. Some of them get loaded in delayed fashion.

Hrm... here's why that's weird though.  The SerializableOptionSet is initially created from teh option service, and it is given the entire set of 'serializable option keys' (and hte values for them).  They are currently 1:1.  So even if someone tried to add a new key later, it will be presumed to be non-serializable since the serializable set is defined at construction point of hte SerializableOptionSet.

Reply:
You might be correct, all the laziness might be coming for new language being loaded for the first time in a solution: https://github.com/dotnet/roslyn/blob/27613724bd239da606824c33e94f52085993c041/src/Workspaces/Core/Portable/Options/GlobalOptionService.cs#L75-L104

Reply:
Yeah, and lazy language loading definitely makes sense as we don't want to load/execute anything in another language dll until necessary (or major regressions up teh wazoooo).  But the option-values seem to have no lazyness in them (which i think is also likely a good thing).  We get a snapshot, and it's a real snapshot of data.